### ISSUE TITLE
Bug: singleInstance not working as expected when resolving from nested containers

### ISSUE THREAD

**Issue by lerppana** [ybainier/Hypodermic#47]
I have a problem when resolving service that is meant to be used as singleInstance from nested containers. I have written a test case here:

```
#include <cassert>

#include <Hypodermic/ContainerBuilder.h>

struct interface
{
    virtual bool& get_value() = 0;
};

struct service : interface
{
    bool& get_value() final { return value; }

    bool value{};
};

int main(int args, const char** argv)
{
    auto builder = std::make_shared<Hypodermic::ContainerBuilder>();

    /// CASE1: this works if I resolve from container instead of nested container
    // using resolved_type = service;
    // builder->registerType<service>().singleInstance();

    /// CASE2: this works if I resolve from container instead of nested container
    // using resolved_type = interface;
    // builder->registerType<service>().as<interface>().singleInstance();

    /// CASE3: this works for both nested container and container
    // using resolved_type = interface;
    // builder->registerInstance(std::make_shared<service>()).as<interface>();

    /// CASE4: this works for both nested container and container
    using resolved_type = service;
    builder->registerInstance(std::make_shared<service>());

    auto container = builder->build();

    {
        auto nested_builder = std::make_shared<Hypodermic::ContainerBuilder>();
        auto nested_container = nested_builder->buildNestedContainerFrom(*container);
        auto service0 = nested_container->resolve<resolved_type>();
        service0->get_value() = true;
    }

    {
        auto nested_builder = std::make_shared<Hypodermic::ContainerBuilder>();
        auto nested_container = nested_builder->buildNestedContainerFrom(*container);
        auto service0 = nested_container->resolve<resolved_type>();
        assert(service0->get_value());
    }

    {
        auto service0 = container->resolve<resolved_type>();
        assert(service0->get_value());
    }

    return 0;
}
```

All of these cases work fine when I resolve the `resolved_type` from `container` instead of `nested_container`. However when resolving from `nested_container` cases 1 and 2 fail both of the asserts.

**Comment by ybainier** [comment]
Hello, I have fixed this issue in the commit a9e1b5b86aa248f8e577d454227ee9fb0c570119. I used your unit tests to showcase the issue's resolution. Is that fine for you?

**Comment by lerppana** [comment]
Yes it works great now, thanks a lot!
